src/transformStateWithClones.js

```diff
@@ -7,7 +7,29 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = {};
+
+  Object.assign(currentState, state);
+
+  for (let i = 0; i < actions.length; i++) {
+    let updatedState = {};
+
+    Object.assign(updatedState, currentState);
+
+    if (actions[i].type === 'addProperties') {
+      Object.assign(updatedState, actions[i].extraData);
+    } else if (actions[i].type === 'removeProperties') {
+      actions[i].keysToRemove.forEach((key) => delete updatedState[key]);
+    } else if (actions[i].type === 'clear') {
+      updatedState = {};
+    }
+
+    stateHistory.push(updatedState);
+    currentState = updatedState;
```

Comment:
Reassigning `currentState` to `updatedState` directly may cause issues, especially after a 'clear' action, as it breaks the reference to the original `currentState` object. Consider using `Object.assign(currentState, updatedState);` to ensure `currentState` is correctly updated without losing its reference.